### ISSUE TITLE
Cache Files Rebuilt On Every Startup

### ISSUE THREAD

**Issue by MrSmite** [CraftedMods/nei-recipe-handlers#28]
**Minecraft:** *1.7.10*
**LotR:** *[1.7.10] LOTRMod B33.8*
**NEI:** *NotEnoughItems-1.7.10-1.0.5.120-universal*
**neiLotr:** *neiLotr-0.6.0-ALPHA*
**neiRecipeHandlers:** *neirecipehandlers-202-alpha*
**CodeChicken:** *CodeChickenLib-1.7.10-1.1.3.138-universal*

**Issue:**

Both *itemCache.dat* and *recipeCache.dat* are rebuilt on every startup during the post-init phase when no mods or recipes have been added or removed.

**Notes:**

This bug report is being filed per the mod author's request [here](https://github.com/CraftedMods/nei-recipe-handlers/issues/27#issuecomment-403759772) to separate the issue out from another issue that has been closed.

[Relevant discussion](https://github.com/CraftedMods/nei-recipe-handlers/issues/27#issuecomment-403668181) from previous thread:

> *I noticed with this iteration, the cache files are still regenerated during post-init and itemCache.dat is slightly larger (95 KB) than with the previous mod version (65 KB). The curious thing though is with recipeCache.dat.*
> 
> *When recipeCache.dat is generated, it sits at 15 KB. When MC is finished loading and you spawn into any SP world, before even opening your inventory, it increases to 22 KB (date modified changes) and stays that way. You can "save and exit", create a new world, go back into the existing one or "quit". The file will remain 22 KB until you restart MC where it is regenerated and back to 15 KB.*

A [complete list of mods](https://github.com/CraftedMods/nei-recipe-handlers/issues/27#issue-332646126) can be found in the OP of the previous report. These are the mods in use when I noticed this issue and can serve as reference for which mods add or change recipes.

**Comment by MrSmite** [comment]
Sorry for the late reply,

My modpack isn't a true modpack. It's just mods that I've put together myself and I think some of them have a "no redistribution" license. I'll go ahead and run the procedure again, it will take me a few days though.

**Comment by CraftedMods** [comment]
@MrSmite I changed the cache management system with the newest version of `neiRecipeHandlers`. Please try the current released version to see whether the bug can be reproduced. Note that the file `itemCache.dat` is no longer used - instead a new file `environmentCache.dat` is relevant. It contains a list of all currently installed mods - of these mods are added/removed or updated, the recipe cache will be recreated.

**Comment by MrSmite** [comment]
@CraftedMods 

Thanks for the update, I'll give it a try. Sorry I never posted back but life happened and I haven't had much time to play Minecraft lately, or any other game for that matter.

**Comment by MrSmite** [comment]
This is still an issue with:

Minecraft 1.7.10
neiLotr-1.0.0-ALPHA
neiRecipeHandlers-2.0.2-ALPHA
LOTRMod v34.3

**Comment by MrSmite** [comment]
Sorry, 'edit' buttond doesn't work... but by "this" I meant the rebuilding of the cache files which seems to cause MC to take a long time to start.

**Comment by CraftedMods** [comment]
Is the cache rebuilt evey startup, or only once?

**Comment by MrSmite** [comment]
> Is the cache rebuilt evey startup now, or only once?

Seems to be the same issue we discussed earlier. During PostInit, the cache is rebuilt, cpu is at 100% and it takes over 2 minutes on an i5-7500 to complete.

Nothing on my end has changed except the update to the mods listed above. I just haven't been playing much and I had temporarily disabled LotR mod anyway.

**Comment by CraftedMods** [comment]
That it is once rebuilt after certain changes, is the way the recipe loading works. The alloy forge recipes of the LOTR Mod are hard to process, and thus it can take that much time on some PCs. But the results should be cached and then reused for later startups, so those should be fast then. Could you send me your `neiRecipeHandlers` configuration file?

**Comment by CraftedMods** [comment]
Also, is there a way I could contact you privately (Discord, E-Mail (you could create a temporary address))? Then you could send me your modpack, and I can try it with your setup.

**Comment by MrSmite** [comment]
I uploaded my homemade modpack with configs, send an e-mail to:

145964877925@discardmail.com

and I will reply with the link.

**Comment by MrSmite** [comment]
NOTE: I received your email and sent you the link. Removed my previous post because I can't edit on github and I didn't want the temporary email to be filled in case I need to resend.

**Comment by MrSmite** [comment]
I was reading past comments here and I wonder if this is still an issue with how MineTweaker exposes custom recipes. I only have a few custom recipes (less than 10) so I can't see why but I also didn't decompile everything to look at it in Eclipse either.

**Comment by CraftedMods** [comment]
Hm, I don't know. I'll try it with and without it, let's see what that changes...

**Comment by MrSmite** [comment]
I'm sure it's obvious from my modpack that Forge is required. I didn't include it because I didn't look up their policy... but i'm using **1.7.10-Forge10.13.4.1614-1.7.10**

**Comment by CraftedMods** [comment]
That's the version I use too.

**Comment by MrSmite** [comment]
Well, I just tried loading by removing MineTweaker but there was no improvement. I remember last time we did this troubleshooting, DynamicRecipes and CraftTweaker were the problem... not so lucky this time ;(

**Comment by CraftedMods** [comment]
I'll be able to do some testing when I receive the E-Mail...

**Comment by MrSmite** [comment]
Tried sending it through a different service: guerrilla mail. Should show up as 'sharklasers.com'

**Comment by MrSmite** [comment]
How about now?

**Comment by CraftedMods** [comment]
Still nothing.

**Comment by MrSmite** [comment]
*sigh*

Sent you the link using my email. Apparently free services are only good for receiving and watching web ads.

**Comment by CraftedMods** [comment]
Ok, received it. I'll do some testing tomorrow. Thanks!

**Comment by CraftedMods** [comment]
Seems like `neiRecipeHandlers` assumes that some mod versions did change, or mods did get added or removed.
Inspecting further...

**Comment by MrSmite** [comment]
I was coming to let you know I noticed the same thing....

My world kept crashing every time I tried to interact with a specific NPC near my spawn. Since it's an old world but I hadn't really done anything in it yet, I decided to "recreate world". I deleted all the NEI cache and it did take a while the first time but now it only takes a few seconds.

Maybe my world was so broken that it was actually the cause. Incidentally that NPC doesn't crash anymore either.

**Comment by MrSmite** [comment]
Thanks though for your prompt attention to this issue.

**Comment by CraftedMods** [comment]
No problem.

**Comment by MrSmite** [comment]
Hmm, after all that I had to revert back to

Minecraft 1.7.10
neiLotr-0.6.0-ALPHA
neiRecipeHandlers-2.0.2-ALPHA
LOTRMod B33.8

There is an incompatibility with LotR and Growthcraft and Waila causing an unrecoverable crash at world load ('Already Tesselating') when standing near a fence with rope attached. 

Disabling Waila stops the crash but the fence / rope is untextured and right-clicking it gives you one of the LotR faction crafting tables...

**Comment by MrSmite** [comment]
*edit*

I know this isn't your bug, just wanted to share in case other people update to LotR 34.3 and crash thinking it's related to your nei plugins.
